aten/src/ATen/native/AveragePool3d.cpp

```diff
@@ -0,0 +1,488 @@
+          int64_t iwidth,
+          int64_t iheight,
+          int64_t otime,
+          int64_t owidth,
+          int64_t oheight,
+          int kT,
+          int kW,
+          int kH,
+          int dT,
+          int dW,
+          int dH,
+          int padT,
+          int padW,
+          int padH,
+          bool count_include_pad)
+{
+  at::parallel_for(0, nslices, 0, [&](int64_t start, int64_t end) {
+    for (auto k = start; k < end; k++)
+    {
+      int64_t i, j, ti;
+
+      /* local pointers. */
+      scalar_t *ip = input_p + k * itime * iwidth * iheight;
+      scalar_t *op = output_p + k * otime * owidth * oheight;
+      for (i = 0; i < otime * oheight * owidth; ++i)
+        *(op + i) = 0;
+
+      /* loop over output */
+      for (ti = 0; ti < otime; ti++)
+      {
+        for (i = 0; i < oheight; i++)
+        {
+          for (j = 0; j < owidth; j++)
+          {
+            /* compute pool range. */
+            int64_t tstart = ti * dT - padT;
+            int64_t hstart = i  * dH - padH;
+            int64_t wstart = j  * dW - padW;
+            int64_t tend = std::min(tstart + kT, itime + padT);
+            int64_t hend = std::min(hstart + kH, iheight + padH);
+            int64_t wend = std::min(wstart + kW, iwidth + padW);
+            int64_t pool_size = (tend - tstart) * (hend - hstart) * (wend - wstart);
+            tstart = std::max(tstart, (int64_t) 0);
+            hstart = std::max(hstart, (int64_t) 0);
+            wstart = std::max(wstart, (int64_t) 0);
+            tend = std::min(tend, itime);
+            hend = std::min(hend, iheight);
+            wend = std::min(wend, iwidth);
+
+            int divide_factor;
+            if (count_include_pad)
+              divide_factor = pool_size;
+            else
+              divide_factor = (tend - tstart) * (hend - hstart) * (wend - wstart);
+
+            /* compute local sum: */
+            scalar_t sum = 0.0;
+            int64_t x, y, z;
+
+            for (z = tstart; z < tend; z++)
+            {
+              for (y = hstart; y < hend; y++)
+              {
+                for (x = wstart; x < wend; x++)
+                {
+                  sum +=  *(ip + z * iwidth * iheight + y * iwidth + x);
+                }
+              }
+            }
+
+            /* set output to local max */
+            *op++ += sum / divide_factor;
+          }
+        }
+      }
+    }
+  });
+}
+
+void avg_pool3d_out_cpu_template(
+  Tensor& output,
+  const Tensor& input_,
+  IntArrayRef kernel_size,
+  IntArrayRef stride,
+  IntArrayRef padding,
+  bool ceil_mode,
+  bool count_include_pad)
+{
+  // #20866 [JIT] stride.empty() is passed through
+  // #20866 [LIBTORCH] IntegrationTest.MNIST: padding.size() == 1
+  TORCH_INTERNAL_ASSERT(kernel_size.size() == 3 &&
+                        (stride.empty() || stride.size() == 3) &&
+                        (padding.size() == 1 || padding.size() == 3),
+    "avg_pool3d: all IntArrayRef sizes must be 3");
```

Comment:
Huh, do we actually check this from the C++ frontend? If not, this probably should be a CHECK.

Reply:
@ezyang Good question, it actually came up now in #22032, which showed that people are using the C++ frontend like the Python functions (expecting that single integers are repeated automatically).

So it's a CHECK now (#22075).